1. Podział na grupy

2. Zrozumienie i analiza problemu, który prezentują dane
• Zapoznanie się z wybranym zestawem danych i opisem zadania do wykonania
• Pobranie bazy danych na swój komputer
• Zdefiniowanie co wiemy na dany temat 
• Efektem tego punktu powinien być krótki 1-2 stronicowy wstęp teoretyczny analizujący problem od strony biznesowej/merytorycznej

In [8]:
#libraries import
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
#%matplotlib inline
import seaborn as sns
import plotnine as pn
#import math as match
#import random as md

In [9]:
#import data
epi = pd.read_csv(r"C:\Users\krzys\OneDrive\Desktop\SDA\91_Projekt_praktyczny_Regresja\epi_r.csv")

In [6]:
#data view - what is included in file?
epi.head()

title  rating  calories  protein  \
0              Lentil, Apple, and Turkey Wrap    2.500     426.0     30.0   
1  Boudin Blanc Terrine with Red Onion Confit    4.375     403.0     18.0   
2                Potato and Fennel Soup Hodge    3.750     165.0      6.0   
3             Mahi-Mahi in Tomato Olive Sauce    5.000       NaN      NaN   
4                    Spinach Noodle Casserole    3.125     547.0     20.0   

    fat  sodium  #cakeweek  #wasteless  22-minute meals  3-ingredient recipes  \
0   7.0   559.0        0.0         0.0              0.0                   0.0   
1  23.0  1439.0        0.0         0.0              0.0                   0.0   
2   7.0   165.0        0.0         0.0              0.0                   0.0   
3   NaN     NaN        0.0         0.0              0.0                   0.0   
4  32.0   452.0        0.0         0.0              0.0                   0.0   

   ...  yellow squash  yogurt  yonkers  yuca  zucchini  cookbooks  leftovers  \
0  ...            0.0     0.0      0.0   0.0       0.0        0.0        0.0   
1  ...            0.0     0.0      0.0   0.0       0.0        0.0        0.0   
2  ...            0.0     0.0      0.0   0.0       0.0        0.0        0.0   
3  ...            0.0     0.0      0.0   0.0       0.0        0.0        0.0   
4  ...            0.0     0.0      0.0   0.0       0.0        0.0        0.0   

   snack  snack week  turkey  
0    0.0         0.0     1.0  
1    0.0         0.0     0.0  
2    0.0         0.0     0.0  
3    0.0         0.0     0.0  
4    0.0         0.0     0.0  

[5 rows x 680 columns]

In [10]:
#the file include many columns which are not visible. To check
# how many is empty; what are details, translation, what is the matrix size
epi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20052 entries, 0 to 20051
Columns: 680 entries, title to turkey
dtypes: float64(679), object(1)
memory usage: 104.0+ MB


In [14]:
column_names = epi.columns.tolist()
for i in range(0, len(column_names), 8):
    print(column_names[i:i+8])

['title', 'rating', 'calories', 'protein', 'fat', 'sodium', '#cakeweek', '#wasteless']
['22-minute meals', '3-ingredient recipes', '30 days of groceries', 'advance prep required', 'alabama', 'alaska', 'alcoholic', 'almond']
['amaretto', 'anchovy', 'anise', 'anniversary', 'anthony bourdain', 'aperitif', 'appetizer', 'apple']
['apple juice', 'apricot', 'arizona', 'artichoke', 'arugula', 'asian pear', 'asparagus', 'aspen']
['atlanta', 'australia', 'avocado', 'back to school', 'backyard bbq', 'bacon', 'bake', 'banana']
['barley', 'basil', 'bass', 'bastille day', 'bean', 'beef', 'beef rib', 'beef shank']
['beef tenderloin', 'beer', 'beet', 'bell pepper', 'berry', 'beverly hills', 'birthday', 'biscuit']
['bitters', 'blackberry', 'blender', 'blue cheese', 'blueberry', 'boil', 'bok choy', 'bon appétit']
['bon app��tit', 'boston', 'bourbon', 'braise', 'bran', 'brandy', 'bread', 'breadcrumbs']
['breakfast', 'brie', 'brine', 'brisket', 'broccoli', 'broccoli rabe', 'broil', 'brooklyn']
['brown ric

In [15]:
#any_null = epi.isnull().any(axis=0)
#print(any_null)

title         False
rating        False
calories       True
protein        True
fat            True
              ...  
cookbooks     False
leftovers     False
snack         False
snack week    False
turkey        False
Length: 680, dtype: bool


In [17]:
#all_zeros = epi.eq(0).all(axis=0)
#print(all_zeros)

title         False
rating        False
calories      False
protein       False
fat           False
              ...  
cookbooks     False
leftovers     False
snack         False
snack week    False
turkey        False
Length: 680, dtype: bool


3. Wstępne zapoznanie się z przedstawioną bazą danych
   • Zapoznanie się z pobraną bazą danych (podłączenie się do niej przez Pythona z użyciem znanych pakietów pandas, numpy)
   • Przygotowanie danych, a w tym:
        * Podział na cechy numeryczne, kategoryczne
        * Przygotowanie danych (pozbycie się lub transformacja wartości niekompletnych)
        * Agregacja niezbędnych informacji
        * Czyszczenie danych
        * Transformacja, dyskretyzacja, skalowanie, grupowanie itp. – użyj to, czego dotyczy Twoje zagadnienie
    • Wyświetlenie podstawowych statystyk (średnia, odchylenie standardowe, mediana) przydatnych do analizy problemu

4. Podsumowanie kroku 2. i 3. oraz dokładne opisanie i przedstawienie hipotez, które będą rozwiązywane
    • Zebranie wyników z punktu 3. i wiedzy z punktu 2.
    • Opis tego, co zamierzamy zrobić wykorzystując zdobyte informacje
    • Opis czego nie możemy zrobić na danej bazie danych z uwagi np. na brak informacji, zbyt dużo wartości odstających
    • Podanie kilku algorytmów/modeli uczenia maszynowego rozwiązujących problemy regresji, który będziemy chcieli użyć w tym projekcie

5. Analiza, wizualizacja danych oraz przeprowadzenie analizy statystycznej w języku Python
Skoro już wiemy co będziemy analizować musimy dokładniej opisać dane za pomocą:
• Wykresów:
    ✔ Wykresy liniowe, histogramy opisujące zależności między pojedynczymi zmiennymi
    ✔ Wykresy opisujące zależności między zmiennymi (Q-Q plot, wykres punktów, mapy ciepła, wykresy słupkowe)
    ✔ Wykresy zmienności w czasie (wykres liniowy)
    ✔ Zaawansowana wizualizacja z wykorzystaniem plotly lub seaborn

6. Użycie i przetestowanie różnych algorytmów regresji do analizy problemów liniowych i nieliniowych
• Wykorzystując wszystkie informacje dotyczące danego problemu, które zgromadziliśmy, a szczególnie opis danych z poprzedniego punktu, zaproponować 3 algorytmy które będą implementowane
• Implementacja algorytmów regresji liniowej (1 i wielu zmiennych), wielomianowej oraz drzewa decyzyjnego
• Zadbanie o minimalizację funkcji kosztu, jeżeli trzeba to z użyciem gradient descent lub innej wbudowanej funkcji w pakiecie sklearn
• Wyświetlenie wartości funkcji kosztu na wykresie dla 10 i 20 iteracji oraz wag algorytmu
• Predykcja algorytmu dla 3 losowych wartości spoza zbioru danych

7A. Ocena działania algorytmu
• Podział danych na zbiór treningowy i walidacyjny
• Zastosowanie walidacji krzyżowej i leave-one-out (przetasować zbiór tak, aby 
zawsze zaczynać od innego miejsca w danych)
• Ocenić skuteczność modeli (współczynnik determinacji, 2 metody błędu 
prognozy)
• Rozwiązanie problemu z przetrenowaniem lub niedotrenowaniem algorytmu
• Znalezienie złotego środka między obciążeniem a wariancją, najlepiej 
pokazując to na wykresie

7B. Wyjaśnienie dlaczego użyty został akurat 
ten algorytm
• Wyciągając wnioski z poprzednich punktów opowiedzieć maksymalnie 
na 1-2 slajdach dlaczego użyliśmy akurat tego modelu, mając na 
uwadze:
- problemy liniowe i nieliniowe
- wpływ statystyk takich jak korelacja na zmienne
- wartości odstające
- wagi algorytmu
- ocena skuteczności jego działania
- zdolności predykcyjne

8. Propozycja jego użycia
• Maksymalnie na 1-2 slajdach odpowiedzieć na pytania:
    Jaki problem jesteśmy w stanie rozwiązać używając naszego algorytmu?
    Jak możemy go użyć? Z użyciem jakich narzędzi przekazywać komuś do 
    użycia?
    Jak interpretować wyniki algorytmu?
    Jak interpretować wyniki skuteczności algorytmu?
    Jak możemy go ulepszyć w przyszłości?

PODSUMOWANIE
1. Określenie trudności podczas pracy w zespole wirtualnym.
2. Określenie, która część sprawiła najwięcej problemów.
3. Wskazanie tematów, które wymagają uzupełnienia lub powtórzenia zagadnień.
4. Czy sposób przeprowadzenia projektu rozwija uczestników i sprawia, że lepiej rozumieją problem, który poznawali na zajęciach?
5. Czy sposób przeprowadzenia projektu może być pomocny przy przyszłych zagadnieniach

Ad 1. Problemy komunikacyjne zwiazane z komunikacją asynchroniczną, Czyli odczytanie inaczej intencji autora
Ad 2. Plik epik to macierz o bardzo dużych rozmiarach (680 na 20058). Sprawia to problem z wyświetleniem danych
    Nie było jasne co należy zrobić z danymi -> jaki wynik uzyskać.
Ad 3. GIT. 
Ad 4. Pytanie retoryczne.
Ad 5. Pytanie retoryczne.

